src/lib/containers/provenance/ProvenanceGraphUtils.ts

```diff
@@ -0,0 +1,270 @@
+    },
+  }
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: entityHeaderIsCurrent,
+  }
+  addNodeAndEdge({
+    newNodeProps: expandNodeProps,
+    existingNodeProps: entityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addEntityPlaceholderNode = (params: {
+  ref: Reference
+  activity: Activity
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { ref, activity, nodesCopy, edgesCopy } = params
+  const entityPlaceholderNodeProps = {
+    type: NodeType.ENTITY_PLACEHOLDER,
+    data: ref,
+  }
+  const activityNodeProps = {
+    type: NodeType.ACTIVITY,
+    data: activity,
+  }
+  addNodeAndEdge({
+    newNodeProps: entityPlaceholderNodeProps,
+    existingNodeProps: activityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addUndefinedNode = (params: {
+  entityHeader: EntityHeader
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { entityHeader, nodesCopy, edgesCopy } = params
+  const undefinedNodeProps = {
+    type: NodeType.UNDEFINED,
+    data: entityHeader,
+  }
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: {
+      entityHeader,
+    },
+  }
+  addNodeAndEdge({
+    newNodeProps: undefinedNodeProps,
+    existingNodeProps: entityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addExternalNode = (params: {
+  usedURL: UsedURL
+  activity: Activity
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { usedURL, activity, nodesCopy, edgesCopy } = params
+  const activityNodeProps = {
+    type: NodeType.ACTIVITY,
+    data: activity,
+  }
+  const externalNodeProps = {
+    type: NodeType.EXTERNAL,
+    data: usedURL,
+  }
+  addNodeAndEdge({
+    newNodeProps: externalNodeProps,
+    existingNodeProps: activityNodeProps,
+    nodesCopy,
+    edgesCopy,
+  })
+}
+
+export const addEntityNode = (params: {
+  entityHeaderIsCurrent: EntityHeaderIsCurrent
+  activity: Activity | undefined
+  nodesCopy: Node[]
+  edgesCopy: Edge[]
+}) => {
+  const { entityHeaderIsCurrent, activity, nodesCopy, edgesCopy } = params
+  const entityNodeProps = {
+    type: NodeType.ENTITY,
+    data: entityHeaderIsCurrent,
+  }
+  if (isNodeNotFound(entityNodeProps, nodesCopy)) {
+    // add the new entity node
+    nodesCopy.push(getProvenanceNode(entityNodeProps))
+    if (activity) {
+      const activityNodeProps = {
+        type: NodeType.ACTIVITY,
+        data: activity,
+      }
+      edgesCopy.push(getProvenanceEdge(entityNodeProps, activityNodeProps))
+    }
+  }
+}
+
+export const getEntityHeadersIsCurrent = async (params: {
+  refs: ReferenceList
+}): Promise<EntityHeaderIsCurrent[]> => {
+  const { refs } = params
+  const usedEntityHeadersPage = await SynapseClient.getEntityHeaders(refs)
+  const refsWithoutVersion: ReferenceList = refs.map(ref => {
+    return { targetId: ref.targetId }
+  })
+  const latestEntityHeadersPage = await SynapseClient.getEntityHeaders(
+    refsWithoutVersion,
+  )
```

Comment:
Still wondering if this call is necessary. EntityHeader has the field `isLatestVersion`. Do you still need to make this request?

Reply:
Mind blown, will modify our EntityHeader type definition and use this new field!